### PR TITLE
Add support BrailleHID  protocol to BrailleEdge device

### PR DESCRIPTION
**Pull request by EdKweon**
### Summary of the issue:
Add support BrailleHID  protocol to BrailleEdge device
### Description of user facing changes
User can use BrailleEdge device using BrailleHID protocol. NVDA keys remain unchanged.
### Description of development approach
Five new keys have been added. On the front edge, starting from the left, there is the leftCursor key, in the center there is the home key, and on the far right there is the rightCursor key. On the top surface, there are control and alt keys on both sides of the spacebar. Additional gesture key mappings can be added to the gesture map based on user demand.
### Testing strategy:
Unit test, Lint test, System test.
### Known issues with pull request:
No.
### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

**Comment by EdKweon** [comment]
NVDA gesture and key remain unchanged. There are no updated in userGuide. 
Thanks!

File: source/brailleDisplayDrivers/hims.py [hunk 1]
```diff
@@ -237,9 +282,12 @@ def __init__(self, port="auto"):
 		for match in self._getTryPorts(port):
 			portType, portId, port, portInfo = match
 			self.isBulk = portType == bdDetect.DeviceType.CUSTOM
+			self.isHID = portType == bdDetect.DeviceType.HID
 			# Try talking to the display.
 			try:
-				if self.isBulk:
+				if self.isHID:
```
Comment (LeonarddeR) [inline source/brailleDisplayDrivers/hims.py:288]: Please consider using a match statement, like this

```suggestion
				match portType:
					case bdDetect.DeviceType.HID:
```

**Review by LeonarddeR — COMMENTED** [review]
I assume this is a custom HID protocol, i.e. not the generic protocol as used by the hidBrailleStandard driver?

File: source/brailleDisplayDrivers/hims.py (hunk 1 above)
Comment (EdKweon) [reply source/brailleDisplayDrivers/hims.py]: Thanks. I converted it with match case.

**Comment by EdKweon** [comment]
I have a question with Lint test. How can I avoid 'line too long' error? 
					case _:
						self._dev = hwIo.Serial(port, baudrate=BAUD_RATE, parity=PARITY, timeout=self.timeout, writeTimeout=self.timeout, onReceive=self._onReceive)

**Comment by EdKweon** [comment]
@LeonarddeR 
> I assume this is a custom HID protocol, i.e. not the generic protocol as used by the hidBrailleStandard driver?

We use hidBrailleStandard driver in device. Is there any reason you think so?

**Comment by LeonarddeR** [comment]
> We use hidBrailleStandard driver in device.

In that case, what is the point of changing the hims driver?

**Comment by EdKweon** [comment]
> > We use hidBrailleStandard driver in device.
> 
> In that case, what is the point of changing the hims driver?


The intention is to support HID rather than to change it. There are users who want HID support. Our device driver testing has been completed on iPhones and MacBooks, so validation for the current device driver is complete. 
If I have misunderstood or provided a different response, please let me know, and I would appreciate the opportunity to correct it. Thank you.

File: source/brailleDisplayDrivers/hims.py [hunk 2]
```diff
@@ -701,7 +799,7 @@ class KeyInputGesture(braille.BrailleDisplayGesture, brailleInput.BrailleInputGe
 
 	source = BrailleDisplayDriver.name
 
-	def __init__(self, model, keys):
+	def __init__(self, model, keys, isHid=False):
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/hims.py:802]: Please add type hints where possible

```suggestion
	def __init__(self, model, keys, isHid: bool = False):
```

File: source/brailleDisplayDrivers/hims.py [hunk 3]
```diff
@@ -236,14 +281,22 @@ def __init__(self, port="auto"):
...
-					self._dev = hwIo.Serial(port, baudrate=BAUD_RATE, parity=PARITY, timeout=self.timeout, writeTimeout=self.timeout, onReceive=self._onReceive)
+				match portType:
+					case bdDetect.DeviceType.HID:
+						self._dev = hwIo.Hid(port, onReceive=self._hidOnReceive)
+					case bdDetect.DeviceType.CUSTOM:
+						# onReceiveSize based on max packet size according to USB endpoint information.
+						self._dev = hwIo.Bulk(port, 0, 1, self._onReceive, onReceiveSize=64)
+					case _:
+						self._dev = hwIo.Serial(
+							port, baudrate=BAUD_RATE,
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/hims.py:294]:
```suggestion
							port,
							baudrate=BAUD_RATE,
```

File: source/brailleDisplayDrivers/hims.py [hunk 4]
```diff
@@ -278,11 +331,26 @@ def __init__(self, port="auto"):
...
 
 	def _sendCellCountRequest(self):
-		log.debug("Sending cell count request...")
-		self._sendPacket(b"\xfb", b"\x01", bytes(32)) # send 32 null bytes
+		if self.isHID:
+			try:
+				data: bytes = self._dev.getFeature(HR_CAPS)
+				self.numCells = data[9]
+			except WindowsError:
+				return  # Fail!
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/hims.py:350]: please log something here, and make the comment more descriptive if necessary

```suggestion
				log.exception("Failed to fetch number of cells")
				return
```

File: user_docs/en/changes.t2t [hunk 5]
```diff
@@ -3,6 +3,12 @@ What's New in NVDA
 
 %!includeconf: ../changes.t2tconf
 
+= 2024.3 =
+
+== New Features ==
+- Added support for the BrailleEdgeS3 braille device(#16279, @EdKweon).
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:9]: the 2024.3 release cycle has not started yet. please move this to the other change log entry.

File: source/brailleDisplayDrivers/hims.py (hunk 4 above)
Comment (EdKweon) [reply source/brailleDisplayDrivers/hims.py]: I think that log is enough. Thanks!

File: user_docs/en/changes.t2t (hunk 5 above)
Comment (EdKweon) [reply user_docs/en/changes.t2t]: Ok. I moved that to 2024.2.

File: user_docs/en/changes.t2t [hunk 6]
```diff
@@ -42,7 +36,7 @@ What's New in NVDA
   - Allows splitting NVDA sounds in one channel (e.g. left) while sounds from all other applications in the other channel (e.g. right).
   - Toggled by ``NVDA+alt+s``.
   -
--
+- - Added support for the BrailleEdgeS3 braille device(#16279, @EdKweon).
```
Comment (seanbudd) [inline user_docs/en/changes.t2t:39]: I would suggest amending the line "Added support for the BrailleEdgeS2 braille device." to "Added support for the BrailleEdgeS2 and BraillEdgeS3 braille devices." and adding both PR numbers

**Review by seanbudd — APPROVED** [review]

File: user_docs/en/changes.t2t (hunk 6 above)
Comment (seanbudd) [reply user_docs/en/changes.t2t:39]: Once that's done this can be merged

Comment (EdKweon) [reply user_docs/en/changes.t2t:39]: I ammended PR. Thanks!

**Comment by EdKweon** [comment]
@seanbudd 
I'm really sorry for not check about error before merge. There are error in class with non defined class member variable. I made a another pull reqeust about this issue.

**Pull request merged**